### PR TITLE
python3.pkgs.numba: report a valid version

### PR DESCRIPTION
**Pull request by Mic92**
just now if packages check for a numba bigger than a specific version they will fail to find this package

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by natsukium** [comment]
Result of `nixpkgs-review pr 248316` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages marked as broken and skipped:</summary>
  <ul>
    <li>piper-train</li>
    <li>piper-train.dist</li>
    <li>python310Packages.graspologic</li>
    <li>python310Packages.graspologic.dist</li>
    <li>python310Packages.qiskit</li>
    <li>python310Packages.qiskit-machine-learning</li>
    <li>python310Packages.qiskit-machine-learning.dist</li>
    <li>python310Packages.qiskit.dist</li>
    <li>python310Packages.scikit-tda</li>
    <li>python310Packages.scikit-tda.dist</li>
  </ul>
</details>
<details>
  <summary>59 packages failed to build:</summary>
  <ul>
    <li>pentestgpt</li>
    <li>pentestgpt.dist</li>
    <li>pianotrans</li>
    <li>pianotrans.dist</li>
    <li>pitivi</li>
    <li>python310Packages.awkward</li>
    <li>python310Packages.awkward.dist</li>
    <li>python310Packages.bambi</li>
    <li>python310Packages.bambi.dist</li>
    <li>python310Packages.clifford</li>
    <li>python310Packages.clifford.dist</li>
    <li>python310Packages.coffea</li>
    <li>python310Packages.coffea.dist</li>
    <li>python310Packages.dask-awkward</li>
    <li>python310Packages.dask-awkward.dist</li>
    <li>python310Packages.dask-glm</li>
    <li>python310Packages.dask-glm.dist</li>
    <li>python310Packages.dask-ml</li>
    <li>python310Packages.dask-ml.dist</li>
    <li>python310Packages.datashader</li>
    <li>python310Packages.datashader.dist</li>
    <li>python310Packages.gumath</li>
    <li>python310Packages.gumath.dist</li>
    <li>python310Packages.hyppo</li>
    <li>python310Packages.hyppo.dist</li>
    <li>python310Packages.langchain</li>
    <li>python310Packages.langchain.dist</li>
    <li>python310Packages.librosa</li>
    <li>python310Packages.librosa.dist</li>
    <li>python310Packages.mplhep</li>
    <li>python310Packages.mplhep.dist</li>
    <li>python310Packages.numbaWithCuda</li>
    <li>python310Packages.numbaWithCuda.dist</li>
    <li>python310Packages.piano-transcription-inference</li>
    <li>python310Packages.piano-transcription-inference.dist</li>
    <li>python310Packages.pymc</li>
    <li>python310Packages.pymc.dist</li>
    <li>python310Packages.pynndescent</li>
    <li>python310Packages.pynndescent.dist</li>
    <li>python310Packages.pyscf</li>
    <li>python310Packages.pyscf.dist</li>
    <li>python310Packages.pytensor</li>
    <li>python310Packages.pytensor.dist</li>
    <li>python310Packages.sparse</li>
    <li>python310Packages.sparse.dist</li>
    <li>python310Packages.stumpy</li>
    <li>python310Packages.stumpy.dist</li>
    <li>python310Packages.torchlibrosa</li>
    <li>python310Packages.torchlibrosa.dist</li>
    <li>python310Packages.trackpy</li>
    <li>python310Packages.trackpy.dist</li>
    <li>python310Packages.umap-learn</li>
    <li>python310Packages.umap-learn.dist</li>
    <li>python310Packages.uproot</li>
    <li>python310Packages.uproot.dist</li>
    <li>python310Packages.vector</li>
    <li>python310Packages.vector.dist</li>
    <li>tts</li>
    <li>tts.dist</li>
  </ul>
</details>
<details>
  <summary>35 packages built:</summary>
  <ul>
    <li>openai-whisper (python310Packages.openai-whisper)</li>
    <li>openai-whisper.dist (python310Packages.openai-whisper.dist)</li>
    <li>python310Packages.aeppl</li>
    <li>python310Packages.aeppl.dist</li>
    <li>python310Packages.aesara</li>
    <li>python310Packages.aesara.dist</li>
    <li>python310Packages.arviz</li>
    <li>python310Packages.arviz.dist</li>
    <li>python310Packages.cppe</li>
    <li>python310Packages.cppe.dist</li>
    <li>python310Packages.ffcv</li>
    <li>python310Packages.ffcv.dist</li>
    <li>python310Packages.galois</li>
    <li>python310Packages.galois.dist</li>
    <li>python310Packages.numba</li>
    <li>python310Packages.numba-scipy</li>
    <li>python310Packages.numba-scipy.dist</li>
    <li>python310Packages.numba.dist</li>
    <li>python310Packages.phik</li>
    <li>python310Packages.phik.dist</li>
    <li>python310Packages.pygmo</li>
    <li>python310Packages.pymatting</li>
    <li>python310Packages.pymatting.dist</li>
    <li>python310Packages.pymoo</li>
    <li>python310Packages.pymoo.dist</li>
    <li>python310Packages.resampy</li>
    <li>python310Packages.resampy.dist</li>
    <li>python310Packages.shap</li>
    <li>python310Packages.shap.dist</li>
    <li>python310Packages.stytra</li>
    <li>python310Packages.stytra.dist</li>
    <li>python310Packages.timezonefinder</li>
    <li>python310Packages.timezonefinder.dist</li>
    <li>python310Packages.xarray-einstats</li>
    <li>python310Packages.xarray-einstats.dist</li>
  </ul>
</details>

**Review by natsukium — APPROVED** [review]
Many packages are broken due to updates of scipy, numpy, and dask.
I think this PR is a better step.

**Comment by natsukium** [comment]
The current patch seems to leave `numba.__version__` at `0+unknown`.
Would you please add another patch?
https://github.com/numba/numba/blob/46059957ad416e68476d1e5f32ccd59f7d5df2bb/numba/_version.py#L639
Since python310Packages.awkward looks at this variable, it cannot determine the numba version.

**Comment by khaneliman** [comment]
Looks like this might resolve https://github.com/NixOS/nixpkgs/issues/248363

**Review by khaneliman — APPROVED** [review]

File: pkgs/development/python-modules/numba/default.nix [hunk 1]
```diff
@@ -77,6 +78,11 @@ in buildPythonPackage rec {
       cuda_toolkit_lib_path = cudatoolkit.lib;
     })
   ];
+  # with untagged version we need to specify the correct version ourselves
+
+  postPatch = ''
+    substituteInPlace setup.py --replace "version=versioneer.get_version()" "version='0.57.1'"
```
Comment (veprbl) [inline pkgs/development/python-modules/numba/default.nix:84]:
```suggestion
    substituteInPlace setup.py --replace "version=versioneer.get_version()" "version='0.57.1'"
    substituteInPlace numba/_version.py \
      --replace 'git_refnames = ""' 'git_refnames = "0.57.1"'
```

Comment (Mic92) [reply pkgs/development/python-modules/numba/default.nix:84]: This file does not exists here: https://github.com/numba/numba

Comment (Mic92) [reply pkgs/development/python-modules/numba/default.nix:84]: Honestly I think the structure of this automatically generated file is more likely to change than the fact that there is a version=versioneer.get_version() in this one.

Comment (veprbl) [reply pkgs/development/python-modules/numba/default.nix:84]: The proposed patch fixes python3Packages.awkward and such

Comment (Mic92) [reply pkgs/development/python-modules/numba/default.nix:84]: I misread your patch actually. I thought it was only patching `numba/_version.py`

Comment (veprbl) [reply pkgs/development/python-modules/numba/default.nix:84]: Oh, yes. I was contemplating proposing that, but in the end also decided that patching both files is more bulletproof.

**Pull request merged**